src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediatePreparedStatement.java

```diff
@@ -107,20 +110,27 @@ public IntermediatePortalStatement bind(
   }
 
   @Override
-  public DescribeMetadata describe() throws Exception {
-    /* Currently, Spanner does not support description of prepared statements which may return
-        result sets (that is to say, SELECT statements). We could return here the describe only for
-        non-SELECT statements, but due to the difficulty in determining the statement type in proxy-
-        side, as well as the unlikelihood of this method being used in production, we just return
-        an exception. If at any point we are able to determine statement types, we can simply return
-        "new DescribeStatementMetadata(
-            Converter.getParams(
-                this.parameterDataTypes,
-                this.totalParameters)
-            );"
-        for non-SELECTs.
-    */
-    throw new IllegalStateException(
-        "Describe statements for non-bound statements are not allowed in spanner.");
+  public DescribeMetadata describe() {
```

Comment:
@pratickchokhani The Golang driver automatically uses prepared statements for all parameterized statements. This is where that ends up in PGAdapter. The 'implementation' in this PR is the simplest possible, and only intended to verify that parameterized statements can be used with the Golang driver, but it can also be a useful starting point for your current investigation.

What basically happens here is that:
1. The user application executes something like `conn.QueryRow(ctx, "SELECT * FROM FOO WHERE BAR=$1", "baz")` in the Golang application (see https://github.com/GoogleCloudPlatform/pgadapter/blob/4a64448855a626ceed7c9585cd6cc5b56feca1a3/src/test/golang/pgx.go#L87).
2. The Golang driver sends a `DescribeStatement` message to PGAdapter with only the SQL string `SELECT * FROM FOO WHERE BAR=$1` and no parameter values. It expects PGAdapter to respond with a `RowDescription` message that contains the metadata of the `ResultSet` that would be returned, and a list of the data types of the parameters in the SQL string.
3. PGAdapter currently responds by executing an `ExecuteSqlRequest` with `QueryMode.PLAN` on Cloud Spanner, and then returning the metadata that is returned by that statement as the `RowDescription`. In addition it just scans through the SQL string to count the number of parameters, and returns an array containing a zero for each parameter it found in the SQL string. Zero in this case means 'the data type of the parameter is unspecified'.